src/mlpack/methods/decision_tree/decision_tree_impl.hpp

```diff
@@ -56,6 +56,98 @@ DecisionTree<FitnessFunction,
   Train(data, labels, numClasses, minimumLeafSize);
 }
 
+//! Construct and train.
+template<typename FitnessFunction,
+         template<typename> class NumericSplitType,
+         template<typename> class CategoricalSplitType,
+         typename ElemType,
+         bool NoRecursion>
+template<typename MatType>
+DecisionTree<FitnessFunction,
+             NumericSplitType,
+             CategoricalSplitType,
+             ElemType,
+             NoRecursion>::DecisionTree(MatType&& data,
+                                        const data::DatasetInfo& datasetInfo,
+                                        arma::Row<size_t>&& labels,
+                                        const size_t numClasses,
+                                        const size_t minimumLeafSize)
+{
+  // move the data to avoid copy
+  MatType tmpData = std::move(data);
+  arma::Row<size_t> tmpLabels = std::move(labels);
```

Comment:
The `std::vector<>` example you showed looks like it is working fine to me.

I see the problem with `MatType` being deduced as `mat&`.  We could use `std::remove_reference` here:

```
typedef std::remove_reference_t<MatType> TrueMatType;
TrueMatType tmpData(std::forward(data));
```

What do you think?